### DISCUSSION TITLE
Files are always served with "Content-Disposition: attachment" - is it desired?

### DISCUSSION THREAD

**Discussion by Lex-2008** [pocketbase/pocketbase#164] [Polls]
I'm not 100% sure what's the desired behaviour here is, so making it a poll.

Currently, files retrieved via files API (` http://localhost:8090/api/files/...`) are always sent with "Content-Disposition: attachment" header - meaning that if you paste URL directly into your browser's address bar - the file will be downloaded by your browser instead of being shown inline (more on this header [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition)) - even if it's a jpeg image or an HTML page.

That makes perfect sense from security POV: if a malicious user uploads an HTML file and tricks a victim user (or, worse, an admin) to open it - they can steal that user's or admin's authorization.

On the other side, it's a bit annoying if you're making a "photo gallery" site - you can't open photo URLs to see them directly - at best, you have to wrap URL into `data:text/html,<img src="...">`.

What would be the best way of doing it in pocketbase?

**Comment by ganigeorgiev** [comment]
As you have already noted, this was done intentional for security reasons, to avoid accidentally loading malicious content in the user browser (its not just html, svgs are also very hard to "purify").

Although, I understand that this may be annoying for some users, I don't think adding an option to change this behavior would be very useful. If someone really wants to change it , they can always put nginx/apache in front of the application.

I'm more eager to make an exception for some common image files with combination of a reasonable default CSP.

**Reply by Lex-2008** [reply]
oh, right, svg, that awesome vector graphics format, which is also HTML.

sounds reasonable! Another workarounds that come to mind are serving the `pb_data/storage` directory with another http server (and rewriting URLs client-side) or using s3 server and serving files from there.

**Reply by ganigeorgiev** [reply]
All file requests are now served with a default CSP - `"default-src 'none'; style-src 'unsafe-inline'; sandbox"`, which should prevent scripts execution even if for some reason the browser decide to render attachment files (inline css is allowed because browsers often inject their own styles for the content preview).

Files with the following mime-types will be served with `Content-Disposition: inline`:
```
  "image/png", "image/jpg", "image/jpeg", "image/gif",
  "video/mp4", "video/3gpp", "video/quicktime", "  video/x-ms-wmv",
```
Everything else will have a `Content-Disposition: attachment`.

The mime types are resolved by [google/go-cloud file attrs on upload, which internally uses `http.DetectContentType`](https://github.com/google/go-cloud/blob/e0f202b696de4fb8dd94ba0149d792456d053c3b/blob/blob.go#L433) by reading the first couple of bytes, so the check is considered safe/trusted.

By default we are using the same mime type from the above check when setting the `Content-Type` header.
Unfortunately, Go strictly follows https://mimesniff.spec.whatwg.org/ and the spec doesn't define svg, so the result content type for svg by default is `text/xml` or similar.
When you directly open the svg url, this doesn't matter too much because the browser will be instructed to download the file,
but this doesn't work well when the svg is loaded as img tag (browsers will not render them if they are not served as `image/svg+xml`).

In order to allow users to load svg as `<img src="test.svg">` (which should be safe), a small exception was made for this case
and if the file has `.svg` extension we will always return a response with `Content-Type: image/svg+xml`.

The changes are now available in [v0.3.2 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.3.2) (together with some other minor fixes).

**Reply by Lex-2008** [reply]
awesome, thanks!

**Reply by Lex-2008** [reply]
> The mime types are resolved by google/go-cloud file attrs on upload, which internally uses http.DetectContentType by reading the first couple of bytes

Oh wow, so you even detect content-type by yourself, instead of relying on content-type sent by uploader. Nice!

**Reply by ganigeorgiev** [reply]
Yes, when serving the file we rely that go-cloud/blob has stored the correct mime-type.
Additionally, we have our own mime-type validator on user upload to ensure that only the admin specified file field mime types are allowed - https://github.com/pocketbase/pocketbase/blob/master/forms/validators/file.go#L33-L63
